File: VB/SpreadWinDemo/filtering/customfilter.vb
Public Class customfilter

    Public Sub New()

        InitializeComponent()

        ' シートの設定
        InitSheet(FpSpread1.Sheets(0))
    End Sub

    Private Sub InitSheet(sheet As FarPoint.Win.Spread.SheetView)
        ' データ連結
        Dim ds As New DataSet()
        ds.ReadXml(Me.GetType().Assembly.GetManifestResourceStream(System.Reflection.Assembly.GetExecutingAssembly().GetName().Name & ".datanum2.xml"))
        sheet.DataSource = ds

        ' 列幅の設定
        sheet.Columns(0).Width = 45
        sheet.Columns(1).Width = 85
        sheet.Columns(2).Width = 135
        sheet.Columns(3).Width = 80
        sheet.Columns(4).Width = 80
        sheet.Columns(5).Width = 80
        sheet.Columns(6).Width = 80

        sheet.ColumnHeaderAutoTextIndex = 0
        sheet.ColumnHeader.RowCount = 2

        ' フィルターを設定
        Dim hf As FarPoint.Win.Spread.HideRowFilter
        Dim mc As MyCustomFilter
        Dim fcd As FarPoint.Win.Spread.FilterColumnDefinition
        hf = New FarPoint.Win.Spread.HideRowFilter(sheet)
        '非表示フィルタの作成
        mc = New MyCustomFilter()
        'カスタムフィルタを作成
        mc.SheetView = sheet
        'シートの設定
        ' 4列目～7列目にカスタムフィルタを設定
        For i As Integer = 0 To 3
            fcd = New FarPoint.Win.Spread.FilterColumnDefinition(i + 3, FarPoint.Win.Spread.FilterListBehavior.Custom)
            fcd.Filters.Add(mc)
            'カスタムフィルタを定義に追加
            'フィルタ定義を非表示フィルタに追加            
            hf.AddColumn(fcd)
        Next

        'フィルタをシートに設定
        sheet.RowFilter = hf
    End Sub

    <Serializable()> _
    Public Class MyCustomFilter
        Inherits FarPoint.Win.Spread.DefaultFilterItem
        Private sv As FarPoint.Win.Spread.SheetView = Nothing

        Public Sub New()
        End Sub
        Public Overrides ReadOnly Property DisplayName() As String
            Get
                Return "1000～5000"
            End Get
        End Property
        'フィルタに表示する文字列
        Public Overrides WriteOnly Property SheetView() As FarPoint.Win.Spread.SheetView
            Set(value As FarPoint.Win.Spread.SheetView)
                sv = value
            End Set
        End Property
        Private Function IsNumeric(ovalue As Object) As Boolean
            Dim _isNumber As New System.Text.RegularExpressions.Regex("^\-?\d+\.?\d*$")
            Dim m As System.Text.RegularExpressions.Match = _isNumber.Match(Convert.ToString(ovalue))
            Return m.Success
        End Function
        Public Function IsFilteredIn(ovalue As Object) As Boolean
            Dim ret As Boolean = False
            If IsNumeric(ovalue) Then
                If [Double].Parse(Convert.ToString(ovalue)) >= 1000 AndAlso [Double].Parse(Convert.ToString(ovalue)) <= 5000 Then
                    ret = True
                End If
            End If
            Return ret
        End Function
        Public Overrides Function Filter(columnIndex As Integer) As Integer()
            Dim ar As New System.Collections.ArrayList()
            Dim val As Object
            For i As Integer = 0 To sv.RowCount - 1
                val = sv.GetValue(i, columnIndex)
                If IsFilteredIn(val) Then
                    ar.Add(i)
                    '条件に合致する行番号をリストに追加
                End If
            Next
            Return DirectCast(ar.ToArray(GetType(Int32)), Int32())
        End Function
        Public Overrides Function ShowInDropDown(columnIndex As Integer, filteredInRowList As Integer()) As Boolean
            ' filteredInRowList引数は現在、フィルター・イン（フィルターの条件に合致した）行リスト
            ' 条件に合うデータが存在する時のみフィルター要素を表示
            If filteredInRowList Is Nothing Then
                For i As Integer = 0 To sv.RowCount - 1
                    Dim value As Object = sv.GetValue(i, columnIndex)
                    If value IsNot Nothing Then
                        If IsFilteredIn(value) Then
                            Return True
                        End If
                    End If
                Next
            Else
                ' 現在の行リストが条件に合致するか確認
                For i As Integer = 0 To filteredInRowList.Length - 1
                    Dim row As Integer = filteredInRowList(i)
                    Dim value As Object = sv.GetValue(row, columnIndex)
                    If value IsNot Nothing Then
                        If IsFilteredIn(value) Then
                            Return True
                        End If
                    End If
                Next
            End If
            Return False
        End Function
        Public Overrides Function Serialize(w As System.Xml.XmlTextWriter) As Boolean
            w.WriteStartElement("MyCustomFilter")
            MyBase.Serialize(w)
            w.WriteEndElement()
            Return True
        End Function
        Public Overrides Function Deserialize(r As System.Xml.XmlNodeReader) As Boolean
            If r.NodeType = System.Xml.XmlNodeType.Element Then
                If r.Name.Equals("MyCustomFilter") Then
                    MyBase.Deserialize(r)
                End If
            End If
            Return True
        End Function
    End Class
End Class

File: CS/SpreadWinDemoCS/filtering/customfilter.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace SpreadWinDemo.filtering
{
    public partial class customfilter : SpreadWinDemo.DemoBase
    {
        public customfilter()
        {
            InitializeComponent();

            // シートの設定
            InitSheet(fpSpread1.Sheets[0]);
        }

        private void InitSheet(FarPoint.Win.Spread.SheetView sheet)
        {
            //データ連結
            DataSet ds = new DataSet();
            ds.ReadXml(this.GetType().Assembly.GetManifestResourceStream(System.Reflection.Assembly.GetExecutingAssembly().GetName().Name + ".SampleData.datanum2.xml"));
            sheet.DataSource = ds;

            // 列幅の設定
            sheet.Columns[0].Width = 45;
            sheet.Columns[1].Width = 85;
            sheet.Columns[2].Width = 135;
            sheet.Columns[3].Width = 80;
            sheet.Columns[4].Width = 80;
            sheet.Columns[5].Width = 80;
            sheet.Columns[6].Width = 80;


            sheet.ColumnHeaderAutoTextIndex = 0;
            sheet.ColumnHeader.RowCount = 2;

            // フィルターを設定
            FarPoint.Win.Spread.HideRowFilter hf;
            MyCustomFilter mc;
            FarPoint.Win.Spread.FilterColumnDefinition fcd;
            hf = new FarPoint.Win.Spread.HideRowFilter(sheet); //非表示フィルタの作成
            mc = new MyCustomFilter(); //カスタムフィルタを作成
            mc.SheetView = sheet; //シートの設定

            // 4列目～7列目にカスタムフィルタを設定
            for (int i = 0; i < 4; i++)
            {
                fcd = new FarPoint.Win.Spread.FilterColumnDefinition(i + 3, FarPoint.Win.Spread.FilterListBehavior.Custom);
                fcd.Filters.Add(mc); //カスタムフィルタを定義に追加
                hf.AddColumn(fcd); //フィルタ定義を非表示フィルタに追加            
            }

            sheet.RowFilter = hf; //フィルタをシートに設定; 
        }

        [Serializable()]
        public class MyCustomFilter : FarPoint.Win.Spread.DefaultFilterItem
        {
            FarPoint.Win.Spread.SheetView sv = null;

            public MyCustomFilter() { }
            public override string DisplayName
            {
                get { return "1000～5000"; }//フィルタに表示する文字列
            }
            public override FarPoint.Win.Spread.SheetView SheetView
            {
                set { sv = value; }
            }
            private bool IsNumeric(object ovalue)
            {
                System.Text.RegularExpressions.Regex _isNumber = new System.Text.RegularExpressions.Regex(@"^\-?\d+\.?\d*$");
                System.Text.RegularExpressions.Match m = _isNumber.Match(Convert.ToString(ovalue));
                return m.Success;
            }
            public bool IsFilteredIn(object ovalue)
            {
                bool ret = false;
                if (IsNumeric(ovalue))
                {
                    if (Double.Parse(Convert.ToString(ovalue)) >= 1000 && Double.Parse(Convert.ToString(ovalue)) <= 5000)
                        ret = true;
                }
                return ret;
            }
            public override int[] Filter(int columnIndex)
            {
                System.Collections.ArrayList ar = new System.Collections.ArrayList();
                object val;
                for (int i = 0; i < sv.RowCount; i++)
                {
                    val = sv.GetValue(i, columnIndex);
                    if (IsFilteredIn(val))
                        ar.Add(i); //条件に合致する行番号をリストに追加
                }
                return (Int32[])(ar.ToArray(typeof(Int32)));
            }
            public override bool ShowInDropDown(int columnIndex, int[] filteredInRowList)
            {
                // filteredInRowList引数は現在、フィルター・イン（フィルターの条件に合致した）行リスト
                // 条件に合うデータが存在する時のみフィルター要素を表示
                if (filteredInRowList == null)
                {
                    for (int i = 0; i < sv.RowCount; i++)
                    {
                        object value = sv.GetValue(i, columnIndex);
                        if (value != null)
                        {
                            if (IsFilteredIn(value))
                                return true;
                        }
                    }
                }
                else
                {
                    // 現在の行リストが条件に合致するか確認
                    for (int i = 0; i < filteredInRowList.Length; i++)
                    {
                        int row = filteredInRowList[i];
                        object value = sv.GetValue(row, columnIndex);
                        if (value != null)
                        {
                            if (IsFilteredIn(value))
                                return true;
                        }
                    }
                }
                return false;
            }
            public override bool Serialize(System.Xml.XmlTextWriter w)
            {
                w.WriteStartElement("MyCustomFilter");
                base.Serialize(w);
                w.WriteEndElement();
                return true;
            }
            public override bool Deserialize(System.Xml.XmlNodeReader r)
            {
                if (r.NodeType == System.Xml.XmlNodeType.Element)
                {
                    if (r.Name.Equals("MyCustomFilter"))
                        base.Deserialize(r);
                }
                return true;
            }
        }
    }
}
